x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopTrainedModelDeploymentAction.java

```diff
@@ -137,22 +167,26 @@ private void normalUndeploy(Task task, PersistentTasksCustomMetadata.PersistentT
                 }
             }
         );
-
         super.doExecute(task, request, finalListener);
     }
 
-    void waitForTaskRemoved(Set<String> taskIds, StopTrainedModelDeploymentAction.Request request,
+    void waitForTaskRemoved(String modelId,
+                            TrainedModelAllocation trainedModelAllocation,
+                            StopTrainedModelDeploymentAction.Request request,
                             StopTrainedModelDeploymentAction.Response response,
                             ActionListener<StopTrainedModelDeploymentAction.Response> listener) {
-        persistentTasksService.waitForPersistentTasksCondition(persistentTasks ->
-                persistentTasks.findTasks(MlTasks.TRAINED_MODEL_DEPLOYMENT_TASK_NAME, t -> taskIds.contains(t.getId())).isEmpty(),
-            request.getTimeout(), ActionListener.wrap(
-                booleanResponse -> {
-                    listener.onResponse(response);
-                },
+        final Set<String> nodesOfConcern = trainedModelAllocation.getNodeRoutingTable().keySet();
+        client.admin()
+            .cluster()
+            .prepareListTasks(nodesOfConcern.toArray(String[]::new))
+            .setDetailed(true)
+            .setWaitForCompletion(true)
+            .setActions(modelId)
+            .setTimeout(request.getTimeout())
+            .execute(ActionListener.wrap(
+                complete -> listener.onResponse(response),
```

Comment:
Specifically, nodes of concern is interesting, it is possible that a NEW node as added to the cluster while this allocation is being removed. So, it may start running on that node. That is why restricting to the CURRENT nodes is important and also why deleting the whole allocation is important.

Maybe we should add a toplevel state to the allocation to indicate its being deleted and prevent this...

Reply:
Indeed. It seems to me the logic here should be:

1. mark allocation as deleting (or stopping)
2. make sure to prevent new routings on a deleting allocation
3. stop all existing routings after we have ensured new ones cannot start
4. wait until the routing table is empty (or something else equivalent)